Add strange pills, RandomFillSolutionComponent

MisterMecky opened the pull request:
<!-- Please read these guidelines before opening your PR: https://docs.spacestation14.io/en/getting-started/pr-guideline -->
<!-- The text between the arrows are comments - they will not be visible on your PR. -->

## About the PR
<!-- What does it change? What other things could this impact? -->
Adds strange pills, which contains 20u of a randomly picked reagent. These pills will spawn randomly in maintenance closets. 

| Reagent (20u)  | Weight    | 
|--------------|-----------|
|# **Elements**|      | 
| Aluminium| 1      | 
| Carbon| 1      | 
| Chlorine | 1      | 
| Copper | 1      | 
| Fluorine | 1      | 
| Hydrogen | 1      | 
| Iodine | 1      | 
| Lithium | 1      | 
| Mercury | 1      | 
| Potassium | 1      | 
| Phosphorus | 1      | 
| Radium | 1      | 
| Silicon | 1     | 
| Sulfur | 1      | 
| Sodium | 1      | 
|# **Medicines**|      | 
| Ipecac| 3  | 
| Omnizine|2 |  
|Tricordrazine  |3 |
|# **Narcotics**|      | 
|Desoxyephedrine  |3 |
|Ephedrine  |3 |
|SpaceDrugs |5 |
|Nocturine |3 |
|MuteToxin  | 3|
|NorepinephricAcid | 3|
|# **Toxins**|      | 
|ChloralHydrate  |3 |
|Mold | 3|
|Pax  | 3|
|Toxin | 5|

Adds a new component, RandomFillSolutionComponent. Upon MapInit, this component will pick a reagent using a weighted random prototype, then add it to the entities solution.

**Media**
<!-- 
PRs which make ingame changes (adding clothing, items, new features, etc) are required to have media attached that showcase the changes.
Small fixes/refactors are exempt.
Any media may be used in SS14 progress reports, with clear credit given.

If you're unsure whether your PR will require media, ask a maintainer.

Check the box below to confirm that you have in fact seen this (put an X in the brackets, like [X]):
-->
![strangecap](https://user-images.githubusercontent.com/62370007/229344292-0c3bd70f-40b2-4163-b801-e111db9facbc.PNG)


https://user-images.githubusercontent.com/62370007/229344312-79d79ac4-870a-47e8-88cd-54d71c0ca49a.mp4

- [X] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

**Changelog**
<!--
Here you can fill out a changelog that will automatically be added to the game when your PR is merged.

Only put changes that are visible and important to the player on the changelog.

Don't consider the entry type suffix (e.g. add) to be "part" of the sentence:
bad: - add: a new tool for engineers
good: - add: added a new tool for engineers

Putting a name after the :cl: symbol will change the name that shows in the changelog (otherwise it takes your GitHub username)
Like so: :cl: PJB
-->

:cl:
- add: Added strange pills, which contain a random reagent. These can be found in maintenance.

vulppine commented:
floor pills are back baby ‼️ ‼️

deltanedas commented:
youd also have to disable spiking drinks to cheese it

vulppine commented:
> youd also have to disable spiking drinks to cheese it

part of the fun of floor pills is dumping everything into a cup and chugging it

deltanedas commented:
yeah but then funnymen will dump them into a 1u ethanol beaker to separate out the omnizine from the nocturine

mirrorcult commented:
yeah and thats funny

deltanedas commented:
ok new chemistry meta that isnt cookie clicker im sold

mirrorcult commented:
i dont think it'll be meta but no reason it can't be an option for aspiring maintchemists

mirrorcult commented:
that said, might need to have a few more 'useless' chem strange pills

DogZeroX commented:
Stuff like raw material pills

Flareguy commented:
i feel like the table should just be a blacklist instead of a whitelist, part of the excitement in SS13 was being able to get literally ANYTHING from a super death chemical to a lethal dose of meth to 50u of vodka.

MisterMecky commented:
Gonna open this for review, at the moment I'm not sure if having a massive list of solutions that a pill can potentially contain is a good idea? 
- Maybe I can split the list off into its own prototype and reference it. 
- Maybe I should delete the component when its done its thing?
Lemme know what I should do.

With the element pills, I was thinking of making a bottle item that contains 30u of a random element, which would spawn randomly in maint closets. If this gets merged I'll probably open a new pr for that.

0x6273 reviewed (changes requested):
Use file-scoped namespaces for new .cs files

metalgearsloth commented on Content.Server/Chemistry/Components/SolutionManager/RandomFillSolutionComponent.cs:
```diff
@@ -0,0 +1,38 @@
+    public List<RandomSolutionEntry> RandomList = new();
+}
+
+[DataDefinition]
+public struct RandomSolutionEntry
+{
+    /// <summary>
+    ///     Weight for this entry.
+    /// </summary>
+    [DataField("weight")]
+    public float Weight { get; set; }
+
+    /// <summary>
+    ///     Solution to add.
+    /// </summary>
+    [DataField("randreagents")]
+    public Solution RandReagents { get; set; }
+}
```
This should be somewhere more generic.

metalgearsloth commented on Content.Server/Chemistry/Components/SolutionManager/RandomFillSolutionComponent.cs:
```diff
@@ -0,0 +1,38 @@
+/// <summary>
+///     Fills a solution container randomly using a weighted list
+/// </summary>
+[RegisterComponent]
```
Access

MisterMecky replied:
`[RegisterComponent, Access(typeof(SolutionRandomFillSystem))]`
Like this?

metalgearsloth replied:
yeah

metalgearsloth commented on Content.Server/Chemistry/EntitySystems/SolutionRandomFillSystem.cs:
```diff
@@ -0,0 +1,49 @@
+        var target = _solutionsSystem.EnsureSolution(uid, component.Solution);
+        var sumOfWeights = 0;
+
+        foreach (var picked  in component.RandomList)
+        {
+            sumOfWeights += (int) picked.Weight;
+        }
```
This is going to have rounding issues though looks like you picked this from elsewhere.

Realistically RandomSolutionEntry should implement an interface and then a helper method for this be added to randomextensions somewhere.

metalgearsloth commented on Content.Server/Chemistry/EntitySystems/SolutionRandomFillSystem.cs:
```diff
@@ -0,0 +1,49 @@
+        Solution? randSolution = null;
+
+        foreach (var picked in component.RandomList)
+        {
+            sumOfWeights -= (int) picked.Weight;
```
Same here on rounding when it's unneeded.

MisterMecky commented:
I've reworked this to use WeightedRandomPrototype and removed RandomSolutionEntry. May need another review?

Overall its less flexible (Cant set individual quantities, multiple reagents), but it does what it needs too. If someone wants these features they can be added later...

I've set the strange pill reagent quantity to always be 20u to reflect the recent pill changes.

metalgearsloth commented on Content.Server/Chemistry/Components/SolutionManager/RandomFillSolutionComponent.cs:
```diff
@@ -0,0 +1,29 @@
+
+    /// <summary>
+    ///     Weighted random prototype Id. Used to pick reagent.
+    /// </summary>
+    [DataField("weightedRandomId")]
+    public string WeightedRandomId { get; set; } = "default";
```
Needs customtypeserializer and also required arg.

metalgearsloth reviewed (changes requested):
Adding the solution struct was fine as long as it was in shared and reuseable but up to you how you want to do it.

MisterMecky commented:
The pill sprite seems to revert back to its default sprite when out of PVS (or whatever its called), I might remove the random sprites...
